File: modules/quantum-crypto/ios/QuantumCryptoModule.swift
import ExpoModulesCore
import Foundation
import Security
import CommonCrypto

public class QuantumCryptoModule: Module {
  public func definition() -> ModuleDefinition {
    Name("QuantumCrypto")

    AsyncFunction("generateKyberKeyPair") { (promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let result = try self.generateKyberKeyPairNative()
          promise.resolve(result)
        } catch {
          print("Kyber key generation error: \(error)")
          promise.reject("KYBER_KEYGEN_ERROR", "Failed to generate Kyber key pair: \(error.localizedDescription)")
        }
      }
    }

    AsyncFunction("generateDilithiumKeyPair") { (promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let result = try self.generateDilithiumKeyPairNative()
          promise.resolve(result)
        } catch {
          print("Dilithium key generation error: \(error)")
          promise.reject("DILITHIUM_KEYGEN_ERROR", "Failed to generate Dilithium key pair: \(error.localizedDescription)")
        }
      }
    }

    AsyncFunction("kyberEncapsulate") { (publicKeyBase64: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          print("Attempting Kyber encapsulation with public key: \(publicKeyBase64.prefix(20))...")
          
          guard let publicKeyData = Data(base64Encoded: publicKeyBase64) else {
            print("Failed to decode base64 public key")
            throw QuantumCryptoError.invalidPublicKey
          }
          
          print("Public key data length: \(publicKeyData.count), expected: \(OQS_KEM_kyber_768_length_public_key)")
          
          // Validate public key length
          guard publicKeyData.count == OQS_KEM_kyber_768_length_public_key else {
            print("Invalid public key length: \(publicKeyData.count), expected: \(OQS_KEM_kyber_768_length_public_key)")
            throw QuantumCryptoError.invalidPublicKeyLength
          }
          
          let result = try self.kyberEncapsulateNative(publicKey: publicKeyData)
          print("Kyber encapsulation successful")
          promise.resolve(result)
        } catch let error as QuantumCryptoError {
          print("Kyber encapsulation error: \(error)")
          promise.reject("KYBER_ENCAP_ERROR", "Error: \(error.localizedDescription)")
        } catch {
          print("Unexpected Kyber encapsulation error: \(error)")
          promise.reject("KYBER_ENCAP_ERROR", "Error: \(error.localizedDescription)")
        }
      }
    }

    AsyncFunction("kyberDecapsulate") { (ciphertextBase64: String, privateKeyBase64: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          guard let ciphertextData = Data(base64Encoded: ciphertextBase64),
                let privateKeyData = Data(base64Encoded: privateKeyBase64) else {
            throw QuantumCryptoError.invalidInput
          }
          
          // Validate input lengths
          guard ciphertextData.count == OQS_KEM_kyber_768_length_ciphertext else {
            throw QuantumCryptoError.invalidCiphertextLength
          }
          
          guard privateKeyData.count == OQS_KEM_kyber_768_length_secret_key else {
            throw QuantumCryptoError.invalidPrivateKeyLength
          }
          
          let sharedSecret = try self.kyberDecapsulateNative(ciphertext: ciphertextData, privateKey: privateKeyData)
          promise.resolve(sharedSecret.base64EncodedString())
        } catch {
          print("Kyber decapsulation error: \(error)")
          promise.reject("KYBER_DECAP_ERROR", "Failed to perform Kyber decapsulation: \(error.localizedDescription)")
        }
      }
    }

    AsyncFunction("dilithiumSign") { (messageBase64: String, privateKeyBase64: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          guard let messageData = Data(base64Encoded: messageBase64),
                let privateKeyData = Data(base64Encoded: privateKeyBase64) else {
            throw QuantumCryptoError.invalidInput
          }
          
          // Validate private key length
          guard privateKeyData.count == OQS_SIG_dilithium_3_length_secret_key else {
            throw QuantumCryptoError.invalidPrivateKeyLength
          }
          
          let signature = try self.dilithiumSignNative(message: messageData, privateKey: privateKeyData)
          promise.resolve(signature.base64EncodedString())
        } catch {
          print("Dilithium signing error: \(error)")
          promise.reject("DILITHIUM_SIGN_ERROR", "Failed to sign with Dilithium: \(error.localizedDescription)")
        }
      }
    }

    AsyncFunction("dilithiumVerify") { (messageBase64: String, signatureBase64: String, publicKeyBase64: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          guard let messageData = Data(base64Encoded: messageBase64),
                let signatureData = Data(base64Encoded: signatureBase64),
                let publicKeyData = Data(base64Encoded: publicKeyBase64) else {
            throw QuantumCryptoError.invalidInput
          }
          
          // Validate public key length
          guard publicKeyData.count == OQS_SIG_dilithium_3_length_public_key else {
            throw QuantumCryptoError.invalidPublicKeyLength
          }
          
          let isValid = try self.dilithiumVerifyNative(message: messageData, signature: signatureData, publicKey: publicKeyData)
          promise.resolve(isValid)
        } catch {
          print("Dilithium verification error: \(error)")
          promise.reject("DILITHIUM_VERIFY_ERROR", "Failed to verify Dilithium signature: \(error.localizedDescription)")
        }
      }
    }
  }

  // MARK: - Native Implementation Methods
  
  private func generateKyberKeyPairNative() throws -> [String: String] {
    var publicKey = [UInt8](repeating: 0, count: Int(OQS_KEM_kyber_768_length_public_key))
    var privateKey = [UInt8](repeating: 0, count: Int(OQS_KEM_kyber_768_length_secret_key))
    
    let result = OQS_KEM_kyber_768_keypair(&publicKey, &privateKey)
    print("Kyber keypair generation result: \(result)")
    
    guard result == OQS_SUCCESS else {
      throw QuantumCryptoError.keyGenerationFailed
    }
    
    let publicKeyData = Data(publicKey)
    let privateKeyData = Data(privateKey)
    
    print("Generated Kyber keys - Public: \(publicKeyData.count) bytes, Private: \(privateKeyData.count) bytes")
    
    return [
      "publicKey": publicKeyData.base64EncodedString(),
      "privateKey": privateKeyData.base64EncodedString()
    ]
  }
  
  private func generateDilithiumKeyPairNative() throws -> [String: String] {
    var publicKey = [UInt8](repeating: 0, count: Int(OQS_SIG_dilithium_3_length_public_key))
    var privateKey = [UInt8](repeating: 0, count: Int(OQS_SIG_dilithium_3_length_secret_key))
    
    let result = OQS_SIG_dilithium_3_keypair(&publicKey, &privateKey)
    print("Dilithium keypair generation result: \(result)")
    
    guard result == OQS_SUCCESS else {
      throw QuantumCryptoError.keyGenerationFailed
    }
    
    let publicKeyData = Data(publicKey)
    let privateKeyData = Data(privateKey)
    
    print("Generated Dilithium keys - Public: \(publicKeyData.count) bytes, Private: \(privateKeyData.count) bytes")
    
    return [
      "publicKey": publicKeyData.base64EncodedString(),
      "privateKey": privateKeyData.base64EncodedString()
    ]
  }
  
  private func kyberEncapsulateNative(publicKey: Data) throws -> [String: String] {
    print("Starting Kyber encapsulation...")
    
    var ciphertext = [UInt8](repeating: 0, count: Int(OQS_KEM_kyber_768_length_ciphertext))
    var sharedSecret = [UInt8](repeating: 0, count: Int(OQS_KEM_kyber_768_length_shared_secret))
    
    let result = publicKey.withUnsafeBytes { publicKeyBytes in
      guard let publicKeyPtr = publicKeyBytes.bindMemory(to: UInt8.self).baseAddress else {
        print("Failed to get public key pointer")
        return OQS_ERROR
      }
      
      print("Calling OQS_KEM_kyber_768_encaps...")
      let encapResult = OQS_KEM_kyber_768_encaps(&ciphertext, &sharedSecret, publicKeyPtr)
      print("OQS_KEM_kyber_768_encaps result: \(encapResult)")
      return encapResult
    }
    
    guard result == OQS_SUCCESS else {
      print("Kyber encapsulation failed with result: \(result)")
      throw QuantumCryptoError.encapsulationFailed
    }
    
    let ciphertextData = Data(ciphertext)
    let sharedSecretData = Data(sharedSecret)
    
    print("Encapsulation successful - Ciphertext: \(ciphertextData.count) bytes, SharedSecret: \(sharedSecretData.count) bytes")
    
    return [
      "ciphertext": ciphertextData.base64EncodedString(),
      "sharedSecret": sharedSecretData.base64EncodedString()
    ]
  }
  
  private func kyberDecapsulateNative(ciphertext: Data, privateKey: Data) throws -> Data {
    print("Starting Kyber decapsulation...")
    
    var sharedSecret = [UInt8](repeating: 0, count: Int(OQS_KEM_kyber_768_length_shared_secret))
    
    let result = ciphertext.withUnsafeBytes { ciphertextBytes in
      privateKey.withUnsafeBytes { privateKeyBytes in
        guard let ciphertextPtr = ciphertextBytes.bindMemory(to: UInt8.self).baseAddress,
              let privateKeyPtr = privateKeyBytes.bindMemory(to: UInt8.self).baseAddress else {
          print("Failed to get pointers for decapsulation")
          return OQS_ERROR
        }
        
        print("Calling OQS_KEM_kyber_768_decaps...")
        let decapResult = OQS_KEM_kyber_768_decaps(&sharedSecret, ciphertextPtr, privateKeyPtr)
        print("OQS_KEM_kyber_768_decaps result: \(decapResult)")
        return decapResult
      }
    }
    
    guard result == OQS_SUCCESS else {
      print("Kyber decapsulation failed with result: \(result)")
      throw QuantumCryptoError.decapsulationFailed
    }
    
    let sharedSecretData = Data(sharedSecret)
    print("Decapsulation successful - SharedSecret: \(sharedSecretData.count) bytes")
    
    return sharedSecretData
  }
  
  private func dilithiumSignNative(message: Data, privateKey: Data) throws -> Data {
    print("Starting Dilithium signing...")
    
    var signature = [UInt8](repeating: 0, count: Int(OQS_SIG_dilithium_3_length_signature))
    var signatureLength = 0
    
    let result = message.withUnsafeBytes { messageBytes in
      privateKey.withUnsafeBytes { privateKeyBytes in
        guard let messagePtr = messageBytes.bindMemory(to: UInt8.self).baseAddress,
              let privateKeyPtr = privateKeyBytes.bindMemory(to: UInt8.self).baseAddress else {
          print("Failed to get pointers for signing")
          return OQS_ERROR
        }
        
        print("Calling OQS_SIG_dilithium_3_sign...")
        let signResult = OQS_SIG_dilithium_3_sign(&signature, &signatureLength, messagePtr, message.count, privateKeyPtr)
        print("OQS_SIG_dilithium_3_sign result: \(signResult), signature length: \(signatureLength)")
        return signResult
      }
    }
    
    guard result == OQS_SUCCESS else {
      print("Dilithium signing failed with result: \(result)")
      throw QuantumCryptoError.signingFailed
    }
    
    let signatureData = Data(signature.prefix(signatureLength))
    print("Signing successful - Signature: \(signatureData.count) bytes")
    
    return signatureData
  }
  
  private func dilithiumVerifyNative(message: Data, signature: Data, publicKey: Data) throws -> Bool {
    print("Starting Dilithium verification...")
    
    let result = message.withUnsafeBytes { messageBytes in
      signature.withUnsafeBytes { signatureBytes in
        publicKey.withUnsafeBytes { publicKeyBytes in
          guard let messagePtr = messageBytes.bindMemory(to: UInt8.self).baseAddress,
                let signaturePtr = signatureBytes.bindMemory(to: UInt8.self).baseAddress,
                let publicKeyPtr = publicKeyBytes.bindMemory(to: UInt8.self).baseAddress else {
            print("Failed to get pointers for verification")
            return OQS_ERROR
          }
          
          print("Calling OQS_SIG_dilithium_3_verify...")
          let verifyResult = OQS_SIG_dilithium_3_verify(messagePtr, message.count, signaturePtr, signature.count, publicKeyPtr)
          print("OQS_SIG_dilithium_3_verify result: \(verifyResult)")
          return verifyResult
        }
      }
    }
    
    let isValid = result == OQS_SUCCESS
    print("Verification result: \(isValid)")
    
    return isValid
  }
}

enum QuantumCryptoError: Error, LocalizedError {
  case invalidPublicKey
  case invalidInput
  case keyGenerationFailed
  case encapsulationFailed
  case decapsulationFailed
  case signingFailed
  case verificationFailed
  case invalidPublicKeyLength
  case invalidPrivateKeyLength
  case invalidCiphertextLength
  
  var errorDescription: String? {
    switch self {
    case .invalidPublicKey:
      return "Invalid or malformed public key"
    case .invalidInput:
      return "Invalid input parameters"
    case .keyGenerationFailed:
      return "Failed to generate cryptographic keys"
    case .encapsulationFailed:
      return "Failed to perform key encapsulation"
    case .decapsulationFailed:
      return "Failed to perform key decapsulation"
    case .signingFailed:
      return "Failed to create digital signature"
    case .verificationFailed:
      return "Failed to verify digital signature"
    case .invalidPublicKeyLength:
      return "Public key has incorrect length"
    case .invalidPrivateKeyLength:
      return "Private key has incorrect length"
    case .invalidCiphertextLength:
      return "Ciphertext has incorrect length"
    }
  }
}

File: modules/quantum-crypto/android/src/main/java/expo/modules/quantumcrypto/QuantumCryptoModule.kt
package expo.modules.quantumcrypto

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import java.security.Security
import java.security.SecureRandom
import android.util.Base64
import android.util.Log
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.bouncycastle.pqc.crypto.crystals.kyber.*
import org.bouncycastle.pqc.crypto.crystals.dilithium.*
import org.bouncycastle.crypto.AsymmetricCipherKeyPair
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import expo.modules.kotlin.functions.Coroutine

class QuantumCryptoModule : Module() {
  
  companion object {
    private const val TAG = "QuantumCryptoModule"
    
    // Expected key and data lengths for validation
    private const val KYBER_768_PUBLIC_KEY_LENGTH = 1184
    private const val KYBER_768_PRIVATE_KEY_LENGTH = 2400
    private const val KYBER_768_CIPHERTEXT_LENGTH = 1088
    private const val KYBER_768_SHARED_SECRET_LENGTH = 32
    
    private const val DILITHIUM_3_PUBLIC_KEY_LENGTH = 1952
    private const val DILITHIUM_3_PRIVATE_KEY_LENGTH = 4000
    
    init {
      // Ensure BouncyCastle provider is installed
      if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME) == null) {
        Security.addProvider(BouncyCastleProvider())
        Log.d(TAG, "BouncyCastle provider added")
      } else {
        Log.d(TAG, "BouncyCastle provider already available")
      }
    }
  }

  // Store key pairs to avoid reconstruction issues
  private val kyberKeyPairs = mutableMapOf<String, AsymmetricCipherKeyPair>()
  private val dilithiumKeyPairs = mutableMapOf<String, AsymmetricCipherKeyPair>()

  override fun definition() = ModuleDefinition {
    Name("QuantumCrypto")

    AsyncFunction("generateKyberKeyPair") Coroutine { promise: Promise ->
      withContext(Dispatchers.Default) {
        try {
          Log.d(TAG, "Starting Kyber key pair generation...")
          
          val keyPairGenerator = KyberKeyPairGenerator()
          keyPairGenerator.init(KyberKeyGenerationParameters(SecureRandom(), KyberParameters.kyber768))
          
          val keyPair = keyPairGenerator.generateKeyPair()
          val publicKey = (keyPair.public as KyberPublicKeyParameters).encoded
          val privateKey = (keyPair.private as KyberPrivateKeyParameters).encoded
          
          Log.d(TAG, "Kyber key generation successful")
          Log.d(TAG, "Generated Kyber keys - Public: ${publicKey.size} bytes, Private: ${privateKey.size} bytes")
          
          // Store the key pair for later use
          val keyId = Base64.encodeToString(publicKey, Base64.NO_WRAP)
          kyberKeyPairs[keyId] = keyPair
          
          val result = mapOf(
            "publicKey" to Base64.encodeToString(publicKey, Base64.NO_WRAP),
            "privateKey" to Base64.encodeToString(privateKey, Base64.NO_WRAP)
          )
          
          promise.resolve(result)
        } catch (e: Exception) {
          Log.e(TAG, "Kyber key generation failed", e)
          promise.reject("KYBER_KEYGEN_ERROR", "Failed to generate Kyber key pair: ${e.message}", e)
        }
      }
    }

    AsyncFunction("generateDilithiumKeyPair") Coroutine { promise: Promise ->
      withContext(Dispatchers.Default) {
        try {
          Log.d(TAG, "Starting Dilithium key pair generation...")
          
          val keyPairGenerator = DilithiumKeyPairGenerator()
          keyPairGenerator.init(DilithiumKeyGenerationParameters(SecureRandom(), DilithiumParameters.dilithium3))
          
          val keyPair = keyPairGenerator.generateKeyPair()
          val publicKey = (keyPair.public as DilithiumPublicKeyParameters).encoded
          val privateKey = (keyPair.private as DilithiumPrivateKeyParameters).encoded
          
          Log.d(TAG, "Dilithium key generation successful")
          Log.d(TAG, "Generated Dilithium keys - Public: ${publicKey.size} bytes, Private: ${privateKey.size} bytes")
          
          // Store the key pair for later use
          val keyId = Base64.encodeToString(publicKey, Base64.NO_WRAP)
          dilithiumKeyPairs[keyId] = keyPair
          
          val result = mapOf(
            "publicKey" to Base64.encodeToString(publicKey, Base64.NO_WRAP),
            "privateKey" to Base64.encodeToString(privateKey, Base64.NO_WRAP)
          )
          
          promise.resolve(result)
        } catch (e: Exception) {
          Log.e(TAG, "Dilithium key generation failed", e)
          promise.reject("DILITHIUM_KEYGEN_ERROR", "Failed to generate Dilithium key pair: ${e.message}", e)
        }
      }
    }

    AsyncFunction("kyberEncapsulate") Coroutine { publicKeyBase64: String, promise: Promise ->
      withContext(Dispatchers.Default) {
        try {
          Log.d(TAG, "Starting Kyber encapsulation...")
          
          val publicKeyBytes = try {
            Base64.decode(publicKeyBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 public key: ${e.message}")
          }
          
          Log.d(TAG, "Public key data length: ${publicKeyBytes.size}")
          
          val publicKey = try {
            // Try to use stored key pair first
            val storedKeyPair = kyberKeyPairs[publicKeyBase64]
            if (storedKeyPair != null) {
              storedKeyPair.public as KyberPublicKeyParameters
            } else {
              // Reconstruct from bytes - this is a simplified approach
              // In practice, you might need a more sophisticated reconstruction
              KyberPublicKeyParameters(KyberParameters.kyber768, publicKeyBytes)
            }
          } catch (e: Exception) {
            throw IllegalArgumentException("Failed to parse Kyber public key: ${e.message}")
          }
          
          val kemGenerator = KyberKEMGenerator(SecureRandom())
          val kemEncapsulation = kemGenerator.generateEncapsulated(publicKey)
          
          Log.d(TAG, "Kyber encapsulation successful")
          Log.d(TAG, "Ciphertext: ${kemEncapsulation.encapsulation.size} bytes, SharedSecret: ${kemEncapsulation.secret.size} bytes")
          
          val result = mapOf(
            "ciphertext" to Base64.encodeToString(kemEncapsulation.encapsulation, Base64.NO_WRAP),
            "sharedSecret" to Base64.encodeToString(kemEncapsulation.secret, Base64.NO_WRAP)
          )
          
          promise.resolve(result)
        } catch (e: IllegalArgumentException) {
          Log.e(TAG, "Invalid input for Kyber encapsulation", e)
          promise.reject("KYBER_ENCAP_ERROR", "Invalid input: ${e.message}", e)
        } catch (e: Exception) {
          Log.e(TAG, "Kyber encapsulation failed", e)
          promise.reject("KYBER_ENCAP_ERROR", "Failed to perform Kyber encapsulation: ${e.message}", e)
        }
      }
    }

    AsyncFunction("kyberDecapsulate") Coroutine { ciphertextBase64: String, privateKeyBase64: String, promise: Promise ->
      withContext(Dispatchers.Default) {
        try {
          Log.d(TAG, "Starting Kyber decapsulation...")
          
          val ciphertextBytes = try {
            Base64.decode(ciphertextBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 ciphertext: ${e.message}")
          }
          
          val privateKeyBytes = try {
            Base64.decode(privateKeyBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 private key: ${e.message}")
          }
          
          Log.d(TAG, "Ciphertext length: ${ciphertextBytes.size}, Private key length: ${privateKeyBytes.size}")
          
          // Find the corresponding key pair by checking all stored pairs
          var privateKey: KyberPrivateKeyParameters? = null
          for ((publicKeyBase64, keyPair) in kyberKeyPairs) {
            val storedPrivateKey = keyPair.private as KyberPrivateKeyParameters
            if (storedPrivateKey.encoded.contentEquals(privateKeyBytes)) {
              privateKey = storedPrivateKey
              break
            }
          }
          
          if (privateKey == null) {
            throw IllegalArgumentException("Private key not found in stored key pairs")
          }
          
          val kemExtractor = KyberKEMExtractor(privateKey)
          val sharedSecret = kemExtractor.extractSecret(ciphertextBytes)
          
          Log.d(TAG, "Kyber decapsulation successful")
          Log.d(TAG, "SharedSecret: ${sharedSecret.size} bytes")
          
          val result = Base64.encodeToString(sharedSecret, Base64.NO_WRAP)
          promise.resolve(result)
        } catch (e: IllegalArgumentException) {
          Log.e(TAG, "Invalid input for Kyber decapsulation", e)
          promise.reject("KYBER_DECAP_ERROR", "Invalid input: ${e.message}", e)
        } catch (e: Exception) {
          Log.e(TAG, "Kyber decapsulation failed", e)
          promise.reject("KYBER_DECAP_ERROR", "Failed to perform Kyber decapsulation: ${e.message}", e)
        }
      }
    }

    AsyncFunction("dilithiumSign") Coroutine { messageBase64: String, privateKeyBase64: String, promise: Promise ->
      withContext(Dispatchers.Default) {
        try {
          Log.d(TAG, "Starting Dilithium signing...")
          
          val messageBytes = try {
            Base64.decode(messageBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 message: ${e.message}")
          }
          
          val privateKeyBytes = try {
            Base64.decode(privateKeyBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 private key: ${e.message}")
          }
          
          Log.d(TAG, "Message length: ${messageBytes.size}, Private key length: ${privateKeyBytes.size}")
          
          // Find the corresponding key pair by checking all stored pairs
          var privateKey: DilithiumPrivateKeyParameters? = null
          for ((publicKeyBase64, keyPair) in dilithiumKeyPairs) {
            val storedPrivateKey = keyPair.private as DilithiumPrivateKeyParameters
            if (storedPrivateKey.encoded.contentEquals(privateKeyBytes)) {
              privateKey = storedPrivateKey
              break
            }
          }
          
          if (privateKey == null) {
            throw IllegalArgumentException("Private key not found in stored key pairs")
          }
          
          val signer = DilithiumSigner()
          signer.init(true, privateKey)
          val signature = signer.generateSignature(messageBytes)
          
          Log.d(TAG, "Dilithium signing successful")
          Log.d(TAG, "Signature: ${signature.size} bytes")
          
          val result = Base64.encodeToString(signature, Base64.NO_WRAP)
          promise.resolve(result)
        } catch (e: IllegalArgumentException) {
          Log.e(TAG, "Invalid input for Dilithium signing", e)
          promise.reject("DILITHIUM_SIGN_ERROR", "Invalid input: ${e.message}", e)
        } catch (e: Exception) {
          Log.e(TAG, "Dilithium signing failed", e)
          promise.reject("DILITHIUM_SIGN_ERROR", "Failed to sign with Dilithium: ${e.message}", e)
        }
      }
    }

    AsyncFunction("dilithiumVerify") Coroutine { messageBase64: String, signatureBase64: String, publicKeyBase64: String, promise: Promise ->
      withContext(Dispatchers.Default) {
        try {
          Log.d(TAG, "Starting Dilithium verification...")
          
          val messageBytes = try {
            Base64.decode(messageBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 message: ${e.message}")
          }
          
          val signatureBytes = try {
            Base64.decode(signatureBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 signature: ${e.message}")
          }
          
          val publicKeyBytes = try {
            Base64.decode(publicKeyBase64, Base64.NO_WRAP)
          } catch (e: IllegalArgumentException) {
            throw IllegalArgumentException("Invalid base64 public key: ${e.message}")
          }
          
          Log.d(TAG, "Message: ${messageBytes.size} bytes, Signature: ${signatureBytes.size} bytes, Public key: ${publicKeyBytes.size} bytes")
          
          val publicKey = try {
            // Try to use stored key pair first
            val storedKeyPair = dilithiumKeyPairs[publicKeyBase64]
            if (storedKeyPair != null) {
              storedKeyPair.public as DilithiumPublicKeyParameters
            } else {
              // Reconstruct from bytes
              DilithiumPublicKeyParameters(DilithiumParameters.dilithium3, publicKeyBytes)
            }
          } catch (e: Exception) {
            throw IllegalArgumentException("Failed to parse Dilithium public key: ${e.message}")
          }
          
          val verifier = DilithiumSigner()
          verifier.init(false, publicKey)
          val isValid = verifier.verifySignature(messageBytes, signatureBytes)
          
          Log.d(TAG, "Dilithium verification completed")
          Log.d(TAG, "Verification result: $isValid")
          
          promise.resolve(isValid)
        } catch (e: IllegalArgumentException) {
          Log.e(TAG, "Invalid input for Dilithium verification", e)
          promise.reject("DILITHIUM_VERIFY_ERROR", "Invalid input: ${e.message}", e)
        } catch (e: Exception) {
          Log.e(TAG, "Dilithium verification failed", e)
          promise.reject("DILITHIUM_VERIFY_ERROR", "Failed to verify Dilithium signature: ${e.message}", e)
        }
      }
    }
  }
}